### PR TITLE
Add SSL caching share for curl

### LINKED ISSUE #6535: CURLE_BAD_FUNCTION_ARGUMENT when calling HTTPS URLs using curl v>=8.12.1
**Describe the bug**
CURL returns CURLE_BAD_FUNCTION_ARGUMENT when a call is made to any HTTPS URL. This error started appearing after curl versions 8.12, which suggests some of the new changes to how CURL is doing SSL caching needs an update to downstream consumers.

***Exception or Stack Trace***
```
Unhandled exception at 0x00007FFC4AA0933A in azure-sdk-for-cpp-testing.exe: Microsoft C++ exception: Azure::Core::Http::TransportException at memory location 0x0000007F7C7FE430.
```
```
>	libcurl-d.dll!schannel_connect(Curl_cfilter * cf, Curl_easy * data, bool * done) Line 1676	C
 	libcurl-d.dll!schannel_recv(Curl_cfilter * cf, Curl_easy * data, char * buf, unsigned __int64 len, CURLcode * err) Line 2071	C
 	libcurl-d.dll!ssl_cf_recv(Curl_cfilter * cf, Curl_easy * data, char * buf, unsigned __int64 len, CURLcode * err) Line 1530	C
 	libcurl-d.dll!Curl_cf_def_recv(Curl_cfilter * cf, Curl_easy * data, char * buf, unsigned __int64 len, CURLcode * err) Line 112	C
 	libcurl-d.dll!Curl_cf_def_recv(Curl_cfilter * cf, Curl_easy * data, char * buf, unsigned __int64 len, CURLcode * err) Line 112	C
 	libcurl-d.dll!Curl_cf_recv(Curl_easy * data, int num, char * buf, unsigned __int64 len, CURLcode * code) Line 250	C
 	libcurl-d.dll!Curl_conn_recv(Curl_easy * data, int sockindex, char * buf, unsigned __int64 blen, __int64 * n) Line 917	C
 	libcurl-d.dll!curl_easy_recv(void * d, void * buffer, unsigned __int64 buflen, unsigned __int64 * n) Line 1271	C
 	azure-core.dll!Azure::Core::Http::CurlConnection::ReadFromSocket(unsigned char * buffer, unsigned __int64 bufferSize, const Azure::Core::Context & context) Line 1117	C++
 	azure-core.dll!Azure::Core::Http::CurlSession::ReadStatusLineAndHeadersFromRawResponse(const Azure::Core::Context & context, bool reuseInternalBuffer) Line 855	C++
 	azure-core.dll!Azure::Core::Http::CurlSession::Perform(const Azure::Core::Context & context) Line 470	C++
 	azure-core.dll!Azure::Core::Http::CurlTransport::Send(Azure::Core::Http::Request & request, const Azure::Core::Context & context) Line 363	C++
 	azure-sdk-for-cpp-testing.exe!main() Line 15	C++
 	azure-sdk-for-cpp-testing.exe!invoke_main() Line 79	C++
 	azure-sdk-for-cpp-testing.exe!__scrt_common_main_seh() Line 288	C++
 	azure-sdk-for-cpp-testing.exe!__scrt_common_main() Line 331	C++
 	azure-sdk-for-cpp-testing.exe!mainCRTStartup(void * __formal) Line 17	C++
 	kernel32.dll!BaseThreadInitThunk() + 23 bytes	Unknown
 	ntdll.dll!RtlUserThreadStart() + 44 bytes	Unknown
```
**To Reproduce**
```
int main()
{
    std::string url{ "https://www.bing.com" };
    auto http_request = Azure::Core::Http::Request{ Azure::Core::Http::HttpMethod::Get, Azure::Core::Url {url} };

    Azure::Core::Http::CurlTransport transport{};
    auto response = transport.Send(http_request, Azure::Core::Context{});
    auto body_vec = response->ExtractBodyStream()->ReadToEnd();

    std::cout << "Finished" << std::endl;
    return 0;
}
```
with curl version >= 8.12.1
***Code Snippet***
```
for (CURLcode readResult = CURLE_AGAIN; readResult == CURLE_AGAIN;)
{
  readResult = curl_easy_recv(m_handle.get(), buffer, bufferSize, &readBytes);
  switch (readResult)
  {
    case CURLE_AGAIN: {
      // start polling operation
      auto pollUntilSocketIsReady = pollSocketUntilEventOrTimeout(
          context, m_curlSocket, PollSocketDirection::Read, 60000L);

      if (pollUntilSocketIsReady == 0)
      {
        throw TransportException("Timeout waiting for socket to read.");
      }
      else if (pollUntilSocketIsReady < 0)
      { // negative value, error while polling
        throw TransportException("Error while polling for socket ready read");
      }

      // Ready to continue download.
      break;
    }
    case CURLE_OK: {
      break;
    }
    default: {
      // Error reading from socket
      throw TransportException(
          "Error while reading from network socket. CURLE code: " + std::to_string(readResult)
          + ". " + std::string(curl_easy_strerror(readResult)));
    }
  }
}
```
**Expected behavior**
No error

**Screenshots**

**Setup (please complete the following information):**
 - OS: Windows 11 25H2
 - IDE : Microsoft Visual Studio Enterprise 2022 (64-bit) - Current Version 17.13.6
 - Version of the Library used: 1.15

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x ] Bug Description Added
- [x ] Repro Steps Added
- [x ] Setup information Added

### PR DESCRIPTION
**Pull request by sushshring**
This change fixes #6535. In CURL version 8.12, they removed the option to use the default ssl cache, instead the cache share needs to be set on the easy/multi handle. See commits: [disable ssl->state](https://github.com/curl/curl/commit/242a1439e7d8cdb72ae6a2fa2e705e2d9a2b7501#diff-7a8031dd7831547facf7afc280d0d5f7f4ac70d118abde2aa830f02a5f0de1a4) and [add magic checks](https://github.com/curl/curl/commit/3754115750db6324fe1d892fe90c44c4e66a66b6). While I couldn't find any docs for the breaking change, adding a call to `curl_easy_setopt` with a created ssl_share fixes the exception.

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [ ] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core/src/http/curl/curl_connection_private.hpp [hunk 1]
```diff
@@ -142,6 +142,7 @@ namespace Azure { namespace Core {
     class CurlConnection final : public CurlNetworkConnection {
     private:
       Azure::Core::_internal::UniqueHandle<CURL> m_handle;
+      Azure::Core::_internal::UniqueHandle<CURLSH> m_sslShareHandle;
```
Comment (sushshring) [inline sdk/core/azure-core/src/http/curl/curl_connection_private.hpp:145]: Need to add the correct template deleter for type CURLSH

**Comment by sushshring** [comment]
Ready for review. Currently only seeing the CheckFailedCrlValidation, which is also failing on master so doesn't seem related to this change.

**Comment by LarryOsterman** [comment]
This enables a new CURL feature by default, however I'm not seeing any evidence of why we need to enable this feature by default.

Could you please explain in more detail why we need to enable this feature?

**Review by LarryOsterman — CHANGES_REQUESTED** [review]

**Comment by antkmsft** [comment]
I would recommend to also add CHANGELOG.md entry in this PR.

**Review by antkmsft — COMMENTED** [review]
To me, the test coverage you add in this PR looks adequate, so I think it is ok to drop the expected coverage 1% (it is hard and probably to excessive to simulate errors being returned from the curl library itself).

The coverage precentages are defined in ci.yml file in the corresponding directory.

File: sdk/core/azure-core/src/http/curl/curl_connection_private.hpp [hunk 2]
```diff
@@ -42,6 +42,33 @@ namespace Azure { namespace Core {
...
+     * @brief   Unique handle for CURLSHWrapper handles
+     */
+    template <> struct UniqueHandleHelper<CURLSHWrapper>
+    {
+      static void FreeCurlShare(CURLSHWrapper* curl_share)
+      {
+        curl_share_cleanup(curl_share->share_handle);
+        free(curl_share);
+      }
+      using type = _internal::BasicUniqueHandle<CURLSHWrapper, FreeCurlShare>;
```
Comment (antkmsft) [inline sdk/core/azure-core/src/http/curl/curl_connection_private.hpp:70]: I think you either use UniqueHandleHelper with CURLSH + FreeCurlShare() which invokes curl_share_cleanup(), or you make your own wrapper, CURLSHWrapper, which allocates it in the constructor, and calls curl_share_cleanup() in destructor, and then you don't need the UniqueHandleHelper - a regualr unique_ptr will do.

Comment (antkmsft) [reply sdk/core/azure-core/src/http/curl/curl_connection_private.hpp:70]: And then you can do `using UniqueCURLSHHandle = UniqueHandle<CURLSH, FreeCurlShare>;`, and then you declare `UniqueCURLSHHandle m_sslShareHandle;`, so there is no confusion between `void*` for the compiler, because in the typedef you provided the second argument for the UniqueHandle.

**Comment by sushshring** [comment]
@LarryOsterman We noticed the error while making a call to the Azure Attestation SDK. Since Azure SDK for CPP uses Curl version 8.10.1, the bug isn't noticed. Switching to a newer version of curl make this issue start to repro. For the purposes of this issue and PR, I wanted to narrow down the scope to the root-cause of the bug, which is why I just mentioned that making a request to any HTTPS URL reproes the bug. Let me know if you'd like me to include a broader scope in either the PR description or the issue and I can update it that way.

**Comment by sushshring** [comment]
@antkmsft For the changelog it seems it needs a new version number. Should I just add it under 1.16.0-beta.1?

File: sdk/core/azure-core/test/ut/curl_options_test.cpp [hunk 3]
```diff
@@ -365,4 +365,41 @@ namespace Azure { namespace Core { namespace Test {
         0);
   }
 
+  TEST(CurlTransportOptions, disableSslCaching)
+  {
+    Azure::Core::Http::CurlTransportOptions curlOptions;
+    curlOptions.EnableCurlSslCaching = true;
```
Comment (antkmsft) [inline sdk/core/azure-core/test/ut/curl_options_test.cpp:371]: Hmm, the test is called "disableSslCaching", yet we do `curlOptions.EnableCurlSslCaching = true`, is this intended?
BTW, should we have tests for both `curlOptions.EnableCurlSslCaching = true` and `curlOptions.EnableCurlSslCaching = false`?

**Review by antkmsft — APPROVED** [review]
@sushshring, yes, please add under the latest unreleased, i.e. 1.16.0-beta.1.

This probably fits more under a new feature and not a bug fix?
So, something like this, I think:
```
### Features Added

- [[#6535]](https://github.com/Azure/azure-sdk-for-cpp/issues/6535) Enable SSL caching for libcurl transport by default, which is backwards compatible behavior with older libcurl versions, so using the default settings won't result in transport error when using libcurl >= 8.12. The option is controlled by `CurlTransportOptions::EnableCurlSslCaching`, and is on by default.
```
I think there are reasons to look at it as a bugfix for curl >= 8.12 as well, I'd leave it up to you, but I think of it as a new feature also because we're adding new API (new option).
Just think of  that section as our users will read it to get a sense of what is new, and whether they want to upgrade, and also if they have been planning to use newer libcurl version, or have tried to use it in the past, and failed, so they rolled back without filing a bug, and this section will give them an idea that the thing they've seen in the past is fexed now.

I am approving the PR, but I expect my new feedback to be addressed (fixed/replied to if it is not needed), as well as the changelog section added, but in general what I see looks good to me. Thank you!

File: sdk/core/azure-core/test/ut/curl_options_test.cpp (hunk 3 above)
Comment (sushshring) [reply sdk/core/azure-core/test/ut/curl_options_test.cpp:371]: Yes I will update the test. Since EnableCurlSslCaching is default, all other tests validate the standard path. I can add one more test, but it will be a line for line duplicate of some other basic validation test. Let me know

File: sdk/core/azure-core/CHANGELOG.md [hunk 4]
```diff
@@ -4,10 +4,10 @@
 
 ### Features Added
 
+- [[#6535]](https://github.com/Azure/azure-sdk-for-cpp/issues/6535) Enable SSL caching for libcurl transport by default, which is backwards compatible behavior with older libcurl versions, so using the default settings won't result in transport error when using libcurl >= 8.12. The option is controlled by `CurlTransportOptions::EnableCurlSslCaching`, and is on by default.
```
Comment (antkmsft) [inline sdk/core/azure-core/CHANGELOG.md:7]:
```suggestion
- [[#6535]](https://github.com/Azure/azure-sdk-for-cpp/issues/6535) Enable SSL caching for libcurl transport by default, which is backwards compatible behavior with older libcurl versions, so using the default settings won't result in transport error when using libcurl >= 8.12. The option is controlled by `CurlTransportOptions::EnableCurlSslCaching`, and is on by default.

```

**Review by LarryOsterman — APPROVED** [review]

**Review by LarryOsterman — APPROVED** [review]
